src/Main.hs

```diff
@@ -72,26 +73,23 @@ trace (Scene objects lights camera config) x y =
   in
     maybe bgColor (getIntersectionColor ray lights objects) intersectedObject
 
-closestObject :: Ray -> [Object] -> Maybe (Double, Object)
+closestObject :: Ray -> [Object] -> Maybe Intersection
 closestObject ray objects =
   let
-    intersections = mapMaybe intersectLambda objects
+    intersections = mapMaybe (objectIntersection ray) objects
   in
     if null intersections then
       Nothing
     else
-      Just $ minimumBy minimumDefinedByFirst intersections
-  where
-    intersectLambda :: Object -> Maybe (Double, Object)
-    intersectLambda obj@(Object s _) = fmap (\i -> (i, obj)) (rayIntersection ray s)
-    minimumDefinedByFirst :: (Double, Object) -> (Double,Object) -> Ordering
-    minimumDefinedByFirst  x y
-      | fst x < fst y = LT
-      | fst x > fst y = GT
-      | otherwise = EQ
-
-getIntersectionColor :: Ray -> [Light] -> [Object] -> (Double, Object) -> Color
-getIntersectionColor ray lights objects (hitDistance, hitObject) =
+      Just $ minimum intersections
+
+objectIntersection :: Ray -> Object -> Maybe Intersection
+objectIntersection ray obj@(Object s _) =
+  (Intersection obj) <$> rayIntersection ray s
```

Comment:
so the `<$>` operator is the infix version of fmap. https://hackage.haskell.org/package/base-4.8.2.0/docs/Data-Functor.html#v:-60--36--62-

happy to move back to `fmap f a` style though